### ISSUE TITLE
defaultMethodOptions should be selectively applied to CORS OPTIONS method

### ISSUE THREAD

**Issue by whimzyLive** [aws/aws-cdk#8615]
# Tl;DR
`defaultMethodOptions` should exclude options method by default or at least have an option to do that.

# Read More
LambdaRestApi has `defaultMethodOption`, when it is specified, it gets applied to all methods for given api resource. which is awesome but doesn't work when CORS is involved. 

Take this simple example, 
where I want all the methods to require `x-api-key` api key so I specify 
```typescript
defaultMethodOption: {
  apiKeyRequired: true
}
```
On top of this I also want CORS to be enabled on all api resources so in `LambdaRestApi` I also specify this
```typescript

defaultCorsPreflightOptions : {
  ... cors options 
} 
```
 
Because of `defaultCorsPreflightOptions` an additional method will be created for all my resources to support CORS but because `defaultMethodOption` applies on all methods, options will also have `apiKeyRequired` , which it shouldn't.

### Environment

  - **CLI Version      : 1.38**
  - **Framework Version: 1.38**
  - **Node.js Version:**  12 -->
  - **OS               :**
  - **Language (Version):** TypeScript (3.8.3) -->

This is :bug: Bug Report

**Comment by nija-at** [comment]
Thanks for filing this issue @whimzyLive.

The 'OPTIONS' method [created as part of CORS preflight setup](https://github.com/aws/aws-cdk/blob/a186c24918fddc697270b794b6603add5a47e947/packages/%40aws-cdk/aws-apigateway/lib/resource.ts#L292) will have to be adjusted such that only properties in `defaultMethodOptions` that apply in the CORS context are filtered.

For the time being, you should be able to work around this using [escape hatches](https://docs.aws.amazon.com/cdk/latest/guide/cfn_layer.html#cfn_layer_raw).

**Comment by asterikx** [comment]
I stumbled across this too.

For now, I have removed the API key requirement from the RestApi and added it to each resource. This isn't ideal though.

**Comment by whimzyLive** [comment]
@asterikx until issue is resolved correctly, you can use [cdk escape hatches](https://docs.aws.amazon.com/cdk/latest/guide/cfn_layer.html#cfn_layer_raw) as also suggested by @nija-at.

For my case I did 
```ts
    const corsMethods = this.apiV1.methods.filter(
      method => method.httpMethod === 'OPTIONS'
    );
    corsMethods.forEach(method => {
      const cfnMethod = method.node.defaultChild as CfnMethod;
      cfnMethod.addPropertyOverride('ApiKeyRequired', false);
      cfnMethod.addPropertyOverride('AuthorizationType', 'NONE');
      cfnMethod.addPropertyDeletionOverride('AuthorizerId');
    });
```

**Comment by asterikx** [comment]
@whimzyLive thanks for sharing! I was not sure how to this.

**Comment by mhart** [comment]
There are extra issues with this and proxy support AFAICT. For example, using `LambdaRestApi`:

```js
const api = new apigateway.LambdaRestApi(stack, 'DummyApi', {
  handler,
  defaultMethodOptions: { authorizationType: apigateway.AuthorizationType.IAM },
  defaultCorsPreflightOptions: { allowOrigins: apigateway.Cors.ALL_ORIGINS },
})

api.methods.filter(m => m.httpMethod === 'OPTIONS')
  .forEach(m => m.node.defaultChild.addPropertyOverride('AuthorizationType', 'NONE'))
```

This works in so much as the OPTIONS methods are changed correctly to not use auth – but I still get 403s when sending OPTIONS requests – and I believe (although happy to be wrong) it's because the auth on the `ANY` methods from the proxy is overriding the (lack of) auth on the `OPTIONS` methods.

If I do this, for example, then `OPTIONS` requests work correctly (and the `POST` request is auth'ed correctly):

```js
const api = new apigateway.LambdaRestApi(stack, 'DummyApi', {
  handler,
  proxy: false,
  defaultCorsPreflightOptions: { allowOrigins: apigateway.Cors.ALL_ORIGINS },
})
api.root.addMethod('POST', null, { authorizationType: apigateway.AuthorizationType.IAM })
```

**Comment by mhart** [comment]
I guess this is akin to the [`AddDefaultAuthorizerToCorsPreflight`](https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/sam-property-api-apiauth.html#sam-api-apiauth-adddefaultauthorizertocorspreflight) property in SAM

**Comment by mhart** [comment]
Another related issue here (if not exactly the same): https://github.com/aws/aws-cdk/issues/906#issuecomment-619426621

**Comment by whimzyLive** [comment]
@mhart I am using `lambdaRestApi` with above workaround and all options are working fine. Only difference is: I am using it with custom Authorizer instead of IAM Authorizer. If you could share a part of your synthesised template that contains `AWS::ApiGateway::Method` I might be able to help.

**Comment by mhart** [comment]
Like I said, I'm pretty sure the generated CFN is actually right – I mean, it's removed the auth from OPTIONS – the problem is that (I suspect) the auth on the `ANY` method is overriding the non-auth on the `OPTIONS` method.

Anyway, here are the generated `AWS::ApiGateway::Method`s:

```json
{
   "DummyApiproxyOPTIONSC056BEDA": {
      "Type": "AWS::ApiGateway::Method",
      "Properties": {
        "HttpMethod": "OPTIONS",
        "ResourceId": {
          "Ref": "DummyApiproxy55B1C1CD"
        },
        "RestApiId": {
          "Ref": "DummyApi80F1E171"
        },
        "AuthorizationType": "NONE",
        "Integration": {
          "IntegrationResponses": [
            {
              "ResponseParameters": {
                "method.response.header.Access-Control-Allow-Headers": "'Content-Type,X-Amz-Date,Authorization,X-Api-Key,X-Amz-Security-Token,X-Amz-User-Agent'",
                "method.response.header.Access-Control-Allow-Origin": "'*'",
                "method.response.header.Access-Control-Allow-Methods": "'OPTIONS,GET,PUT,POST,DELETE,PATCH,HEAD'"
              },
              "StatusCode": "204"
            }
          ],
          "RequestTemplates": {
            "application/json": "{ statusCode: 200 }"
          },
          "Type": "MOCK"
        },
        "MethodResponses": [
          {
            "ResponseParameters": {
              "method.response.header.Access-Control-Allow-Headers": true,
              "method.response.header.Access-Control-Allow-Origin": true,
              "method.response.header.Access-Control-Allow-Methods": true
            },
            "StatusCode": "204"
          }
        ]
      }
    },
    "DummyApiproxyANY1EBD4910": {
      "Type": "AWS::ApiGateway::Method",
      "Properties": {
        "HttpMethod": "ANY",
        "ResourceId": {
          "Ref": "DummyApiproxy55B1C1CD"
        },
        "RestApiId": {
          "Ref": "DummyApi80F1E171"
        },
        "AuthorizationType": "AWS_IAM",
        "Integration": {
          "IntegrationHttpMethod": "POST",
          "Type": "AWS_PROXY",
          "Uri": {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition"
                },
                ":apigateway:",
                {
                  "Ref": "AWS::Region"
                },
                ":lambda:path/2015-03-31/functions/",
                {
                  "Fn::GetAtt": [
                    "DummyFunction3BB5AE03",
                    "Arn"
                  ]
                },
                "/invocations"
              ]
            ]
          }
        }
      }
    }
}
```

**Comment by whimzyLive** [comment]
Okay I see you are using Lambda proxy integration in particular, For Lambda proxy
`your backend is responsible for returning the Access-Control-Allow-Origin and Access-Control-Allow-Headers headers, because a proxy integration doesn't return an integration response.` - from docs

This is mentioned [here](https://docs.aws.amazon.com/apigateway/latest/developerguide/how-to-cors.html), under *Enabling CORS support for Lambda or HTTP proxy integrations*.

**Comment by mhart** [comment]
That doesn't make sense – the 403 happens before the Lambda is even invoked.

**Comment by whimzyLive** [comment]
This was the last thing I could think of. If you are getting 403 on options itself then we might have had different issue. Let me know if you find out something.

**Comment by mhart** [comment]
Found the issue – it was a complicated scenario with CFN stack updates and CDK deployment hashes not being updated. More info in the updated comment here: https://github.com/aws/aws-cdk/issues/8615#issuecomment-658478450

**Comment by jimfum** [comment]
I also have the above issue and have come up with the same CfnMethod workaround. Couldn't believe there isn't anything better, anything official without Cfn and that maybe `defaultMethodOptions` isn't popular enough(?).

SAM has a solution for it though `AddDefaultAuthorizerToCorsPreflight` (https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/sam-property-api-apiauth.html).

**Comment by leantorres73** [comment]
> defaultMethodOptions

and how did you fix it?

**Comment by leantorres73** [comment]
> I also have the above issue and have come up with the same CfnMethod workaround. Couldn't believe there isn't anything better, anything official without Cfn and that maybe `defaultMethodOptions` isn't popular enough(?).
> 
> SAM has a solution for it though `AddDefaultAuthorizerToCorsPreflight` (https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/sam-property-api-apiauth.html).

and how did you fix it? I'm having issues also

**Comment by jimfum** [comment]
Something like the below in TypeScript. Omitting referenced variables but hope it's clear. 

```typescript
const api = new apiGateway.LambdaRestApi(
  this,
  "ABCD",
  {
    restApiName: apiName,
    handler: mainFunc,
    domainName: {
      domainName: apiDomain,
      certificate,
    },
    defaultCorsPreflightOptions: {
      allowOrigins: corsAllowedOrigins,
    },
    defaultMethodOptions: { authorizer: authorizer },
  }
);

api.methods
  .filter((method) => method.httpMethod === "OPTIONS")
  .forEach((method) => {
    const methodCfn = method.node.defaultChild as apiGateway.CfnMethod;
    methodCfn.authorizationType = apiGateway.AuthorizationType.NONE;
    methodCfn.authorizerId = undefined;
  });

```

**Comment by markcarroll** [comment]
@corymhall please re-open. See comment in https://github.com/aws/aws-cdk/pull/22402#issuecomment-1325436342
